### PR TITLE
UCT: replace MAX_BCOPY by SEG_SIZE attribute

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Replace `MAX_BCOPY` by `SEG_SIZE` attribute that is set directly to tranpsort's segment size

## Why ?

`MAX_BCOPY` is not needed anymore

## How ?

Remove `MAX_BCOPY` and `TM_MAX_BCOPY` attributes and replace them by `SEG_SIZE`

### CODE REVIEW

File: src/uct/base/uct_iface.c [hunk 1]
```diff
@@ -540,10 +540,10 @@ UCS_CONFIG_DEFINE_ARRAY(alloc_methods, sizeof(uct_alloc_method_t),
                         UCS_CONFIG_TYPE_ENUM(uct_alloc_method_names));
 
 ucs_config_field_t uct_iface_config_table[] = {
-  {"MAX_BCOPY", "8192",
+  {"SEG_SIZE", "8192",
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: i think we should remove SEG_SIZE from common code and push it down only to relevant transports only (e.g IB,MM)

Comment (yosefe) [reply src/uct/base/uct_iface.c]: also, need to fix the description

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10001/ for details (Mellanox internal link).

File: src/uct/base/uct_iface.c (hunk 1 above)
Comment (dmitrygladkov) [reply src/uct/base/uct_iface.c]: > i think we should remove SEG_SIZE from common code and push it down only to relevant transports only (e.g IB,MM)

yes, this is good point. done - now this is available for IB, MM, SELF, TCP transports (i.e. only for transports that depend on this value)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10009/ for details (Mellanox internal link).

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 2]
```diff
@@ -11,7 +11,14 @@
 #include <uct/ib/rc/base/rc_iface.h>
 
 ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
-  {"RC_", "", NULL,
+  {"RC_",
+#if IBV_HW_TM
+   /* Maximal size of copy-out sends when tag-matching offload is enabled */
+   "IB_SEG_SIZE=48k",
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: we kept it separate from seg size, so 48K is used when HWTM is really enabled

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -56,6 +56,10 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
   {"", "", NULL,
    ucs_offsetof(uct_ib_iface_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_iface_config_table)},
 
+  {"SEG_SIZE", "8192",
+   "Maximal size of copy-out sends.\n",
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: Why reduced description?

File: src/uct/ib/base/ib_iface.c [hunk 4]
```diff
@@ -56,6 +56,10 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
   {"", "", NULL,
    ucs_offsetof(uct_ib_iface_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_iface_config_table)},
 
+  {"SEG_SIZE", "8192",
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c]: how does it correlate with UCX_SEG_SIZE, used for pre-registered buffers in AM/UCP? If I set UCX_SEG_SIZE to some value, will it affect this var as well?

File: test/gtest/uct/test_many2one_am.cc [hunk 5]
```diff
@@ -24,6 +24,25 @@ class test_many2one_am : public uct_test {
     test_many2one_am() : m_am_count(0) {
     }
 
+    void init() {
+        std::string val = "16k";
+        std::string name;
+
+        if (has_rc_or_dc() || has_ud()) {
```
Comment (brminich) [inline test/gtest/uct/test_many2one_am.cc]: maybe add `has_ib`?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10013/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10015/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@brminich can take a look pls? looks like your comments were fixed

File: src/uct/base/uct_iface.c [hunk 6]
```diff
@@ -540,11 +540,6 @@ UCS_CONFIG_DEFINE_ARRAY(alloc_methods, sizeof(uct_alloc_method_t),
                         UCS_CONFIG_TYPE_ENUM(uct_alloc_method_names));
 
 ucs_config_field_t uct_iface_config_table[] = {
-  {"MAX_BCOPY", "8192",
-   "Maximal size of copy-out sends. The transport is allowed to support any size\n"
-   "up to this limit, the actual size can be lower due to transport constraints.",
-   ucs_offsetof(uct_iface_config_t, max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
-
```
Comment (shamisp) [inline src/uct/base/uct_iface.c]: Since you completely remove this, this will break the tuning scripts that users already had ? I think we have to preserve the legacy config field and chose between two in the runtime.

File: src/uct/ib/base/ib_iface.c (hunk 4 above)
Comment (shamisp) [reply src/uct/ib/base/ib_iface.c]: @dmitrygladkov, the SEG_SIZE also used for BCOPY, right ? If so, I suggest update documentation to reflect this more explicitly. SEG_SIZE is used to XXX protocols to define XXX.

**Comment by yosefe** [comment]
since it's not trivial to add alias across different configuration hierarchy levels, we want to add support for "deprecation notice" for evn. vars. When a user tries to set a deprecated config var, there will be a deprecation warning and a suggestion which new var should be used. Since the configuration is not really part of API, and should be used mainly for debugging purposes, IMHO it's good enough.

**Comment by dmitrygladkov** [comment]
> we want to add support for "deprecation notice" for evn. vars.

@yosefe @shamisp 
Opened #3595 PR that adds `deprecated env vars` infrastructure. It will be used in this PR for `UCX_MAX_BCOPY` then

**Comment by shamisp** [comment]
@yosefe - sounds reasonable.

**Comment by dmitrygladkov** [comment]
this PR has been updated to include changes from #3595 PR. #3595 has been closed as it's not needed anymore

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10254/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10382/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10386/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@brminich could you review pls?

File: src/uct/sm/mm/base/mm_iface.h [hunk 7]
```diff
@@ -32,13 +32,14 @@
 
 
 typedef struct uct_mm_iface_config {
-    uct_sm_iface_config_t        super;
-    unsigned                     fifo_size;      /* Size of the receive FIFO */
-    double                       release_fifo_factor;
-    ucs_ternary_value_t          hugetlb_mode;   /* Enable using huge pages for
+    uct_sm_iface_config_t    super;
+    size_t                   seg_size;             /* Size of the receive descriptor (for payload) */
```
Comment (brminich) [inline src/uct/sm/mm/base/mm_iface.h]: comment alignment

File: test/gtest/ucp/test_ucp_tag_match.cc [hunk 8]
```diff
@@ -18,7 +18,7 @@ class test_ucp_tag_match : public test_ucp_tag {
     {
         m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_ENABLE", "y"));
         if (RUNNING_ON_VALGRIND) {
-            m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_MAX_BCOPY", "8k"));
+            m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_SEG_SIZE", "8k"));
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_match.cc]: you introduced TM_MAX_SIZE name, not TM_SEG_SIZE

File: test/gtest/ucp/test_ucp_tag_xfer.cc [hunk 9]
```diff
@@ -37,7 +37,7 @@ class test_ucp_tag_xfer : public test_ucp_tag {
         modify_config("MAX_RNDV_LANES", "2");
         m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_ENABLE", "y"));
         if (RUNNING_ON_VALGRIND) {
-            m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_MAX_BCOPY", "8k"));
+            m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_SEG_SIZE", "8k"));
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: you introduced TM_MAX_SIZE name, not TM_SEG_SIZE

File: src/uct/ib/base/ib_iface.c (hunk 4 above)
Comment (dmitrygladkov) [reply src/uct/ib/base/ib_iface.c]: @shamisp Thanks for the comment! This is fixed now

File: src/uct/sm/mm/base/mm_iface.h (hunk 7 above)
Comment (dmitrygladkov) [reply src/uct/sm/mm/base/mm_iface.h]: fixed

File: test/gtest/ucp/test_ucp_tag_match.cc (hunk 8 above)
Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_match.cc]: I was wrong with the name of the new introduced attribute

File: test/gtest/ucp/test_ucp_tag_xfer.cc (hunk 9 above)
Comment (dmitrygladkov) [reply test/gtest/ucp/test_ucp_tag_xfer.cc]: fixed the name of the config attribute

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10393/ for details (Mellanox internal link).

**Comment by brminich** [comment]
jenkins issue

bot:mlx:retest

File: src/uct/base/uct_iface.c [hunk 10]
```diff
@@ -545,10 +545,10 @@ ucs_config_field_t uct_iface_config_table[] = {
    "UCX_<IB transport>_TX_MIN_INLINE for IB, UCX_MM_FIFO_SIZE for MM",
    UCS_CONFIG_DEPRECATED_FIELD_OFFSET, UCS_CONFIG_TYPE_DEPRECATED},
 
-  {"MAX_BCOPY", "8192",
-   "Maximal size of copy-out sends. The transport is allowed to support any size\n"
-   "up to this limit, the actual size can be lower due to transport constraints.",
-   ucs_offsetof(uct_iface_config_t, max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
+  {"MAX_BCOPY", "",
+   "The configuration parameter replaced by: "
+   "UCX_<transport>_SEG_SIZE for IB, MM, SELF, TCP",
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: "UCX_<transport>_SEG_SIZE where <transport> is one of: IB, MM, SELF, TCP"

File: src/uct/base/uct_iface.h [hunk 11]
```diff
@@ -286,14 +286,12 @@ typedef struct uct_tl_component {
  * Specific transport extend this structure.
  */
 struct uct_iface_config {
-    size_t            max_bcopy;
-
     struct {
-        uct_alloc_method_t  *methods;
-        unsigned            count;
+        uct_alloc_method_t *methods;
```
Comment (yosefe) [inline src/uct/base/uct_iface.h]: unrelated

File: src/uct/base/uct_iface.h [hunk 12]
```diff
@@ -286,14 +286,12 @@ typedef struct uct_tl_component {
...
+        unsigned           count;
     } alloc_methods;
 
-    int               failure;   /* Level of failure reports */
+    int                    failure;   /* Level of failure reports */
```
Comment (yosefe) [inline src/uct/base/uct_iface.h]: unrelated

File: src/uct/ib/base/ib_iface.c [hunk 13]
```diff
@@ -56,6 +56,11 @@ ucs_config_field_t uct_ib_iface_config_table[] = {
...
+   "Maximal size of copy-out sends. The transport is allowed to support any size\n"
+   "up to this limit, the actual size can be lower due to transport constraints.",
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: this documentation as copied from base is irrelevant.
"Size of bounce buffers used for post_send and post_recv"

File: src/uct/ib/base/ib_iface.h [hunk 14]
```diff
@@ -59,58 +59,60 @@ enum {
...
 
     struct {
-        unsigned            queue_len;       /* Queue length */
-        unsigned            max_batch;       /* How many fragments can be batched to one post send */
-        unsigned            max_poll;        /* How many wcs can be picked when polling tx cq */
-        size_t              min_inline;      /* Inline space to reserve for sends */
-        size_t              inl_resp;        /* Inline space to reserve for responses */
-        unsigned            min_sge;         /* How many SG entries to support */
-        unsigned            cq_moderation;   /* How many TX messages are batched to one CQE */
+        unsigned                 queue_len;       /* Queue length */
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h]: all indentation fixes here are unrelated

File: src/uct/ib/cm/cm_iface.c [hunk 15]
```diff
@@ -287,14 +287,14 @@ static UCS_CLASS_INIT_FUNC(uct_cm_iface_t, uct_md_h md, uct_worker_h worker,
         return UCS_ERR_INVALID_PARAM;
     }
 
-    self->num_outstanding     = 0;
-    self->num_completions     = 0;
-    self->service_id          = 0;
-    self->config.timeout_ms   = (int)(config->timeout * 1e3 + 0.5);
+    self->num_outstanding        = 0;
+    self->num_completions        = 0;
+    self->service_id             = 0;
```
Comment (yosefe) [inline src/uct/ib/cm/cm_iface.c]: all indentation fixes here are unrelated

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 16]
```diff
@@ -35,9 +35,15 @@ ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
...
 
-  {"TM_MAX_BCOPY", "48k",
+  {"TM_SEG_SIZE", "48k",
    "Maximal size of copy-out sends when tag-matching offload is enabled",
-   ucs_offsetof(uct_rc_mlx5_iface_common_config_t, tm.max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
+   ucs_offsetof(uct_rc_mlx5_iface_common_config_t, tm.seg_size),
+   UCS_CONFIG_TYPE_MEMUNITS},
+
+  {"TM_MAX_BCOPY", NULL,
+   "Maximal size of copy-out sends when tag-matching offload is enabled",
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: no need documentation for alias field

File: src/uct/sm/mm/base/mm_iface.c [hunk 17]
```diff
@@ -29,6 +29,10 @@ static ucs_config_field_t uct_mm_iface_config_table[] = {
      "Size of the receive FIFO in the memory-map UCTs.",
      ucs_offsetof(uct_mm_iface_config_t, fifo_size), UCS_CONFIG_TYPE_UINT},
 
+    {"SEG_SIZE", "8k",
+     "Maximal size of the receive descriptor (for payload).",
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_iface.c]: "Size of send/receive buffers for copy-out sends"

File: src/uct/sm/self/self.c [hunk 18]
```diff
@@ -35,6 +35,10 @@ static ucs_config_field_t uct_self_iface_config_table[] = {
      ucs_offsetof(uct_self_iface_config_t, super),
      UCS_CONFIG_TYPE_TABLE(uct_iface_config_table)},
 
+    {"SEG_SIZE", "8k",
+     "Maximal send size.",
```
Comment (yosefe) [inline src/uct/sm/self/self.c]: "Size of copy-out buffer"

File: src/uct/tcp/tcp_iface.c [hunk 19]
```diff
@@ -18,6 +18,10 @@ static ucs_config_field_t uct_tcp_iface_config_table[] = {
    ucs_offsetof(uct_tcp_iface_config_t, super),
    UCS_CONFIG_TYPE_TABLE(uct_iface_config_table)},
 
+  {"SEG_SIZE", "8k",
+   "Maximal send size.",
```
Comment (yosefe) [inline src/uct/tcp/tcp_iface.c]: "Size of copy-out buffer"

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10397/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_iface.c (hunk 19 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp_iface.c]: fixed

File: src/uct/sm/self/self.c (hunk 18 above)
Comment (dmitrygladkov) [reply src/uct/sm/self/self.c]: fixed

File: src/uct/sm/mm/base/mm_iface.c (hunk 17 above)
Comment (dmitrygladkov) [reply src/uct/sm/mm/base/mm_iface.c]: fixed

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 16 above)
Comment (dmitrygladkov) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: fixed

File: src/uct/ib/cm/cm_iface.c (hunk 15 above)
Comment (dmitrygladkov) [reply src/uct/ib/cm/cm_iface.c]: removed

File: src/uct/ib/base/ib_iface.h (hunk 14 above)
Comment (dmitrygladkov) [reply src/uct/ib/base/ib_iface.h]: removed

File: src/uct/ib/base/ib_iface.c (hunk 13 above)
Comment (dmitrygladkov) [reply src/uct/ib/base/ib_iface.c]: fixed

File: src/uct/base/uct_iface.h (hunk 12 above)
Comment (dmitrygladkov) [reply src/uct/base/uct_iface.h]: removed

File: src/uct/base/uct_iface.h (hunk 11 above)
Comment (dmitrygladkov) [reply src/uct/base/uct_iface.h]: removed

File: src/uct/base/uct_iface.c (hunk 10 above)
Comment (dmitrygladkov) [reply src/uct/base/uct_iface.c]: fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10399/ for details (Mellanox internal link).

File: src/uct/base/uct_iface.c [hunk 20]
```diff
@@ -545,10 +545,10 @@ ucs_config_field_t uct_iface_config_table[] = {
    "UCX_<IB transport>_TX_MIN_INLINE for IB, UCX_MM_FIFO_SIZE for MM",
    UCS_CONFIG_DEPRECATED_FIELD_OFFSET, UCS_CONFIG_TYPE_DEPRECATED},
 
-  {"MAX_BCOPY", "8192",
-   "Maximal size of copy-out sends. The transport is allowed to support any size\n"
-   "up to this limit, the actual size can be lower due to transport constraints.",
-   ucs_offsetof(uct_iface_config_t, max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
+  {"MAX_BCOPY", "",
+   "The configuration parameter replaced by: "
+   "UCX__SEG_SIZE where is one of: IB, MM, SELF, TCP",
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: sorry, looks like github removed '<>' from my comment. it should be:
"UCX_<transport>_SEG_SIZE where <transport> is one of: IB, MM, SELF, TCP""

Comment (dmitrygladkov) [reply src/uct/base/uct_iface.c]: Ok, I see. Thanks! Fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10409/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> Test FAILed.
> See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10409/ for details (Mellanox internal link).

```
15:47:09 [  FAILED  ] mm/uct_p2p_rma_test.get_bcopy/3, where GetParam() = mm/posix/loopback
15:47:09 [  FAILED  ] mm/uct_p2p_rma_test.put_bcopy/3, where GetParam() = mm/posix/loopback
```

unrelated to this PR

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10415/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> Test FAILed.
> See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10415/ for details (Mellanox internal link).

related to #3555

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10421/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> Test FAILed.
> See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10421/ for details (Mellanox internal link).

```
22:33:08 INFO: Retrying request to {s}->https://repo.maven.apache.org:443
22:49:28 [INFO] ------------------------------------------------------------------------
22:49:28 [INFO] BUILD FAILURE
22:49:28 [INFO] ------------------------------------------------------------------------
22:49:28 [INFO] Total time: 01:05 h
22:49:28 [INFO] Finished at: 2019-05-27T22:49:28+03:00
22:49:28 [INFO] Final Memory: 27M/1157M
22:49:28 [INFO] ------------------------------------------------------------------------
22:49:28 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-jar-plugin:2.4:jar (default-jar) on project jucx: Execution default-jar of goal org.apache.maven.plugins:maven-jar-plugin:2.4:jar failed: Plugin org.apache.maven.plugins:maven-jar-plugin:2.4 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-plugin-api:jar:2.0.6 from/to central (https://repo.maven.apache.org/maven2): Connection timed out (Read failed) -> [Help 1]
22:49:28 [ERROR] 
22:49:28 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
22:49:28 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
22:49:28 [ERROR] 
22:49:28 [ERROR] For more information about the errors and possible solutions, please read the following articles:
22:49:28 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
22:49:28 make[4]: *** [install-data-hook] Error 1
22:49:28 make[4]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node2/worker/0/build-test/bindings/java/src/main/native'
22:49:28 make[3]: *** [install-data-am] Error 2
22:49:28 make[3]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node2/worker/0/build-test/bindings/java/src/main/native'
22:49:28 make[2]: *** [install-am] Error 2
22:49:28 make[2]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node2/worker/0/build-test/bindings/java/src/main/native'
22:49:28 make[1]: *** [install] Error 2
22:49:28 make[1]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr-3/label/hpc-test-node2/worker/0/build-test/bindings/java/src/main/native'
```

unrelated maven error

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10423/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe ok to merge?
